Draft-Accepted/RFCXXXX-Using-IDisposables.md

```diff
@@ -0,0 +1,399 @@
+The equivalent syntax in C# would be this example:
+
+```csharp
+using (FileStream fs = File.OpenRead(path))
+using (CryptoStream cryptoStream = new CryptoStream(...))
+using (StreamReader sr = new StreamReader(cryptoStream, Encoding.ASCII))
+{
+    sr.ReadToEnd();
+}
+```
+
+## Specification
+
+The syntax for the proposed `Using` statement would be:
+
+```
+using ($<disposable>; ...) {
+    <Statement list>
+}
+```
+
+The part inside the parenthesis is a variable to dispose after the `<Statement list>` ends.
+It can contain multiple variables as well as the code used to define the variable.
+Some examples of this would be:
+
+```powershell
+# Creates the instance separately from the using statement
+$key = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+using ($key) {
+    $key.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+}
+
+# Creates the variable inside the parenthesis
+using ($key = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment') {
+    $key.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+}
+
+# Disposes multiple values at the end of the block
+$machineKey = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+$userKey = Get-Item 'HKCU:\Environment'
+
+using ($machineKey; $userKey) {
+    $m = $machineKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    $u = $userKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    "${m};${u}"
+}
+
+# Same as the above but with the vars defined in the parenthesis
+using (
+    $machineKey = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+    $userKey = Get-Item 'HKCU:\Environment'
+) {
+    $m = $machineKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    $u = $userKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    "${m};${u}"
+}
+```
+
+## Implementation Details
+
+I propose adding a new AST type `UsingDisposableStatementAst` that contains
+
++ `PipelineBaseAst[]` - Disposables
++ `StatementBlockAst` - Body
+
+When visited the Ast will visit the `Disposables` to find any `AssignmentStatementAst` or `VariableExpressionAst` which it can use to keep track of variables to dispose.
+The compiler will then emit an Expression that wraps the `Body` with the below for each `Disposables` (in definition order):
+
+```powershell
+$v = ...  # Value from assignment/variable
+try {
+    ...
+}
+finally {
+    ${v}?.Dispose()
+}
+```
+
+## Alternate Proposals and Considerations
+
+Some questions that would need consensus from the maintainers are:
+
++ The name of the statement
++ Whether to support multiple disposables in one block
++ Should it dispose the variable or actual instance captured before the body is run
++ Support only `IDisposable` vs duck typing with `Dispose()` method
++ What to do in case of an error in the dispose
++ Future support for `using` declarations without the block/braces
```

Comment:
Yea it's definitely not part of this RFC but it is something I brought up to consider as possible future work or something that should at least be considering in case we need to adjust the behaviour here to accomodate that in the future.